46.md

```diff
@@ -0,0 +1,159 @@
+* **Signer**: Nostr app that holds the private key of a nostr account and *can sign* on its behalf.
+
+
+## `TL;DR`
+
+
+**App** and **Signer** sends ephemeral encrypted messages to each other using kind `24133`, using a relay of choice. 
+
+App prompts the Signer to do things such as fetching the public key or signing events.
+
+The `content` field must be an encrypted JSONRPC-ish **request** or **response**.
+
+## Signer Protocol
+
+### Messages
+
+#### Request
+
+```json
+{
+  "id": <random_string>,
+  "method": <one_of_the_methods>,
+  "params": [<anything>, <else>]
+}
+```
+
+#### Response
+
+```json
+{
+  "id": <request_id>,
+  "result": <anything>,
+  "error": <reason>
+}
+```
+
+### Methods
+
+
+#### Mandatory 
+
+These are mandatory methods the remote signer app MUST implement:
+
+- **get_public_key**
+  - params []
+  - result `pubkey` 
+- **sign_event**
+  - params [`event`]
+  - result `signature` 
+
+#### optional
+
+
+- **connect**
+  - params [`pubkey`]
+- **disconnect**
+  - params []
+- **delegate** 
+  - params [`pubkey`, `conditions query string`]
+  - result `nip26 delegation token`
+- **get_relays**
+  - params []
+  - result `{ [url: string]: {read: boolean, write: boolean} }` 
+- **nip04_encrypt**
+  - params [`pubkey`, `plaintext`]
+  - result `nip4 ciphertext`
+- **nip04_decrypt**
+  - params [`pubkey`, `nip4 ciphertext`]
+  - result [`plaintext`]
+
+
+NOTICE: `pubkey` and `signature` are hex-encoded strings.
+
+
+### Nostr Connect URI
+
+**Signer** discovers **App** by scanning a QR code, clicking on a deep link or copy-pasting an URI.
+
+The **App** generates a special URI with prefix `nostrconnect://` and base path the hex-encoded `pubkey` with the following querystring parameters **URL encoded**
+
+- `relay` URL of the relay of choice where the **App** is connected and the **Signer** must send and listen for messages.
+- `metadata`  metadata JSON of the **App** 
+    - `name` human-readable name of the **App** 
+    - `url` (optional) URL of the website requesting the connection
+    - `description` (optional) description of the **App**
+    - `icons` (optional) array of URLs for icons of the **App**.
+
+#### JavaScript
+
+```js
+const uri = `nostrconnect://<pubkey>?relay=${encodeURIComponent("wss://relay.damus.io")}&metadata=${encodeURIComponent(JSON.stringify({"name": "Example"}))}`
+```
+
+#### Example
+```sh
+nostrconnect://foobar?relay=wss%3A%2F%2Frelay.damus.io&metadata=%7B%22name%22%3A%22Example%22%7D
+```
+
+
+
+## Flows
+
+The `content` field contains encrypted message as specified by [NIP04](https://github.com/nostr-protocol/nips/blob/master/04.md). The `kind` chosen is `24133`.
+
+### Connect
+
+1. User clicks on **"Connect"** button on a website or scan it with a QR code
+2. It will show an URI to open a "nostr connect" enabled **Signer** 
+3. In the URI there is a pubkey of the **App** ie. `nostrconnect://<pubkey>&relay=<relay>&metadata=<metadata>`
```

Comment:
I think that it could be very important to sign the provided data `relay` & `metadata` by the corresponding application (add a `signature` field). Maybe it would also make sense to add a `timestamp` to the metadata. Also something like `valid-for` could be used to define how long the generated "URI" will be valid. Also maybe `pubkey` could be removed, because it could be recovered from the given signature.

Reply:
With the last part I'm not sure right now. I just discovered nostr - I did not check yet whether the signature algorithm used here is really supporting public key recovery - but I guess yes ;)

Reply:
>  be signed by the corresponding application 

Be signed by who? The web developer? 

Also, Is the signature hardcoded in the client? At that point, everyone can replicate your website by copying the same signature. How does this improve anything? Remember here webapps are ideally without backend/user storage, as that depends on Nostr entirely

Reply:
> identity of an application is also defined by a private key

But that private key is ephemeral, and identifies the session, not the app. It does not protect anything signing the metadata, since anyone will just do that.

Reply:
Hmm.. ok.. I see it like that the application's identity should also be defined by a keypair - maybe additional to the session.. if you don't see why this might be good, no prob.. but maybe it would be useful to introduce two different kinds of URIs? one simple without a signature, and one with - but signed by the application, so that users can verify it's identity easily..

Reply:
I see the value if the app could have a backend to perform a challenge/response auth.

Let's have an example: I create an app, I create a keypair that identify myself the developer, I make a signature of the URI. Where that signature is going to be stored? Hardcoded in the client right? App is basically open source (thats how browser web works, you download the HTML locally)

What would stop a possible phisher to copy paste my app with the signature included? What that signature was necessary for?


Maybe I am misunderstanding, but I fail to see what an hardcoded signature can be useful for...

Reply:
From my point of view we should not exclude applications here that may have a "backend". 

One use-case would be a simple single-sign-on (e.g. "login with nostr") on arbitrary websites. Here that website was configured to use a specific private key that could be used to generate the URI signature. A user that is interacting with that website would be sure that he is really interacting with exactly this website. 

However, I think in the future there will be mobile wallets, where users can install arbitrary applications or services. Any other user may want to interact with these applications/services in a secure and safe way. For this, the user will need to present the connection URI to other users in a way, that these other users can verify it's integrity. So that the users can be sure that they are really interacting with this particular application/service provided by a specific user.

I think in general it would be very important to not restrict this mechanism to static websites. I think that this simple NIP can be used to not only create connections between a user and a static website, it can be used to connect users with other users in general, where it can also be used to connect users with arbitrary services/applications - I think it makes sense that a user is able to verify the identity of a specific service/applications.

One interesting example would be where an arbitrary service or website is using a service that is defined by the user - e.g. the user wants to offer a database service that can be used by other services/websites/applications. For this a QR code maybe shown by that service/website. The user scans this QR code - it is signed, so the user can be sure that the user interacts with a trustworthy service/application/website. That service/application/website is then able to interact with the service defined by the user. In the context of a simple database service it could be just a simple key-value store. However, the service/application/website itself  is not storing the data in their database, it's using the database of the connected user. The user is under full control of what data the service/application/website is storing.

However, this database example is just an example. But I think it would be very important to ensure that this mechanism is as generic as possible while offering security & safety. The user need to be very sure with whom the user is really interacting. And - from my point of view - this is only possible if the Nostr Connect URI supports signatures.

Reply:
Why not use [Login with Nostr](https://github.com/nostr-protocol/nips/issues/154) then?

We spent quite some time in the last month to figure out to merge the two proposals, but unfortunately not have had that much luck.

that being said, fine for me to have an optional **signature** field in the metadata. Thoughts @giowe @vforvalerio87 ?